File: src/networking/SyncAlgorithm.java
package networking;


import java.util.LinkedList;
import java.util.List;

import networking.Client.RemoteNode;

public abstract class SyncAlgorithm {
	
	protected List<RemoteNode> network;
	protected String ip;
	protected volatile boolean isCalcDone;
	protected volatile boolean isPending;
	private volatile boolean isSessionDone;
	private volatile boolean canStart;

	protected static SyncAlgorithm instance = null;

	public SyncAlgorithm(List<RemoteNode> networkInput, String ip)
	{
		canStart     = false;
		isSessionDone = false;
		isPending = false;
		isCalcDone = true;
		// pay attention to the difference between network and this.network
		this.network = new LinkedList<RemoteNode>();
		this.network.addAll(networkInput);
		this.ip      = ip;
	}
	
	public synchronized boolean isCalcDone() {
		return isCalcDone;
	}
	
	public synchronized void setCalcDone(){
		isCalcDone = true;
	}
	
	public synchronized void setCalcInProgress(){
		isCalcDone = false;
	}
	
	public static SyncAlgorithm getInstance()
	{
		return instance;
	}

	abstract public boolean canAccess();

	public synchronized void setPending() {
		System.out.println("Setting isPending");
		isPending = true;
	}

	public void clearPending() {
		isPending = false;
	}

	public synchronized void setSessionDone() {
		isSessionDone = true;
	}
	
	public synchronized boolean isSessionDone(){
		return isSessionDone;
	}

	public void setCanStart() {
		canStart = true;
	}
	public boolean canStart()
	{
		return canStart;
	}
}

File: CSharp/SyncAlgorithm.cs
using System;
using System.Collections.Generic;

namespace Networking
{

    public abstract class SyncAlgorithm
    {

        protected internal List<RemoteNode> network;
        protected internal string ip;
        protected volatile bool isCalcDone;
        public volatile bool isPending;
        private volatile bool isSessionDone;
        protected internal volatile bool canStart;
        public Object mutualLock;

        protected internal static SyncAlgorithm instance = null;

        public SyncAlgorithm(IList<RemoteNode> network, string ip)
        {

            mutualLock = new Object();
            canStart = false;
            isPending = false;
            isSessionDone = false;
            isCalcDone = true;
            // pay attention to the difference between network and this.network
            this.network = new List<RemoteNode>();
            this.network.AddRange(network);
            this.ip = ip;
        }

        public virtual bool CalcDone
        {
            get
            {
                lock (this)
                {
                    return isCalcDone;
                }
            }
        }

        public virtual void setCalcDone()
        {
            lock (this)
            {
                isCalcDone = true;
            }
        }

        public virtual void setCalcInProgress()
        {
            lock (this)
            {
                isCalcDone = false;
            }
        }

        public static SyncAlgorithm Instance
        {
            get
            {
                return instance;
            }
        }

        abstract public bool canAccess();


        public virtual void setPending()

        {
            lock (this)
            {
                //Console.WriteLine("Setting isPending");
                isPending = true;
            }
        }

        public virtual void setCanStart() 
        {
            lock (this)
            {
                
             canStart = true;
            }
        }

        public bool getcanStart()
        {
            return canStart;
        }


        public virtual void clearPending()
        {
            lock (this)
            {
                isPending = false;
            }
        }

        public void setSessionDone() {
            lock (this)
            {
		        isSessionDone = true;
            }

	    }
        public bool SessionDone(){
            lock (this)
            {
                return isSessionDone;
            }

	    }

    }

}